rslint_core/src/groups/errors/valid_typeof.rs

```diff
@@ -0,0 +1,120 @@
+use crate::rule_prelude::*;
+use ast::{BinExpr, Expr, Literal, UnaryExpr};
+
+declare_lint! {
+    /**
+    Enforce the use of valid string literals in a `typeof` comparison.
+
+    ## Invalid Code Examples
+    ```ignore
+    typeof foo === "strnig"
+    typeof foo == "undefimed"
+    typeof bar != "nunber"
+    typeof bar !== "fucntion"
+    ```
+    */
+    #[derive(Default)]
+    ValidTypeof,
+    errors,
+    "valid-typeof",
+
+    #[serde(default)]
+    pub require_string_literals: bool,
+}
+
+const VALID_TYPES: [&str; 8] = [
+    "undefined",
+    "object",
+    "boolean",
+    "number",
+    "string",
+    "function",
+    "symbol",
+    "bigint",
+];
+
+#[typetag::serde]
+impl CstRule for ValidTypeof {
+    fn check_node(&self, node: &SyntaxNode, ctx: &mut RuleCtx) -> Option<()> {
+        let expr = node.try_to::<BinExpr>()?;
+        if !expr.comparison() {
+            return None;
+        }
+        let (lhs, rhs) = (expr.lhs()?, expr.rhs()?);
+
+        let cmp_value = if is_typeof_expr(&lhs) {
+            rhs
+        } else if is_typeof_expr(&rhs) {
+            lhs
+        } else {
+            return None;
+        };
+
+        let str_literal = cmp_value
+            .syntax()
+            .try_to::<Literal>()
+            .and_then(|lit| Some((lit.inner_string_text()?, lit.range())));
+
+        let (literal, literal_range) = if self.require_string_literals {
+            if let Some(lit) = str_literal {
+                lit
+            } else if is_typeof_expr(&cmp_value) {
+                return None;
+            } else {
+                let err = ctx
+                    .err(self.name(), "invalid typeof comparison value")
+                    .primary(cmp_value.range(), "");
+                ctx.add_err(err);
+                return None;
+            }
+        } else {
+            str_literal?
+        };
+
+        if !VALID_TYPES.iter().any(|ty| *ty == literal) {
+            let literal = String::from(literal);
+            let suggestion =
+                util::find_best_match_for_name(VALID_TYPES.iter().copied(), &literal, None);
+
+            let err = ctx
+                .err(self.name(), "invalid typeof comparison value")
```

Comment:
```suggestion
                .err(self.name(), format!("invalid typeof comparison value `{}`", literal))
```
The error message should be understandable on its own, so i would like to include the invalid value in there too 🙂

Reply:
Should the comparison value also be printed if the value is an arbitrary object?

Reply:
Id say only print it if its smaller than say, 20 characters, so if its too large then that wont spam the user